### DISCUSSION TITLE
De/serialize class implementing IList plus additional properties

### DISCUSSION THREAD

**Discussion by chobo2** [YAXLib/YAXLib#219] [Q&A] [answered]
Hi

I have an xml doc like this

```
<root>
     <item> 
            <title>Empire Burlesque</title> 
        </item> 
        <item> 
            <title>Hide your heart</title> 
        </item> 
</root>
```
```
public class Root 
{
   public List<Item> Item {get; set;}
}

public class Item 
{
  public string title {get; set;}
}
```

How can I deserialize this into a c# object. I tried to make an "Item" class with all of these elements in them and then to my "root" class as a collection but that does not work.

I then tried to put the YAXCollection attribute on the property but that did not work either.

**Comment by axunonb** [comment]
It's always helpful to see the code that you've tried so far, and also to elaborate "did not work": What did you get vs. what did you expect. 

```CSharp
public class Item
{
    public string Name { get; set; }
}

[YAXSerializeAs("root")]
public class AllItems : List<Item>
{}

public void Serialize()
{
    var data = new AllItems(){ new(){Name = "First"}, new(){Name = "Second"} };
    var s = new YAXSerializer<AllItems>();
    var xml = s.Serialize(data);
}
```

**Reply by chobo2** [reply]
I am trying to deserialize not serialize.

here is my complete code

```
string xml  = @"<root>
     <item> 
            <title>Empire Burlesque</title> 
        </item> 
        <item> 
            <title>Hide your heart</title> 
        </item> 
</root>";

YAXSerilaizer ser = new YAXSerilaizer (typeof(Root));
var o = ser.Deserialize(xml);

public class Root 
{
   public List<Item> Item {get; set;}
}

public class Item 
{
  public string title {get; set;}
}

```

this will error out saying "no element title" found.

**Reply by axunonb** [reply]
The code sample is now modified for a copy and paste of your use case 😉
```CSharp
public class item
{
    public string title { get; set; }
}

[YAXSerializeAs("root")]
public class AllItems : List<Item>
{}

public void SerializeAndDeserialize()
{
    var data = new AllItems(){ new(){title = "First"}, new(){title = "Second"} };
    var s = new YAXSerializer<AllItems>();
    var xml = s.Serialize(data);
    /* xml is:
<root>
  <item>
    <title>First</title>
  </item>
  <item>
    <title>Second</title>
  </item>
</root>
     */
    var allItems = s.Deserialize(xml);
}
```

**Reply by chobo2** [reply]
> The code sample is now modified for a copy and paste of your use case 😉
> 
> ```cs
> public class item
> {
>     public string title { get; set; }
> }
> 
> [YAXSerializeAs("root")]
> public class AllItems : List<Item>
> {}
> 
> public void SerializeAndDeserialize()
> {
>     var data = new AllItems(){ new(){title = "First"}, new(){title = "Second"} };
>     var s = new YAXSerializer<AllItems>();
>     var xml = s.Serialize(data);
>     /* xml is:
> <root>
>   <item>
>     <title>First</title>
>   </item>
>   <item>
>     <title>Second</title>
>   </item>
> </root>
>      */
>     var allItems = s.Deserialize(xml);
> }
> ```

Thanks, this does seem to work, though on my real implementation it is not working, so I got to see what the differences are.

What happens if I need to do multiple ones, like you I see your using inheritance but can only do it once.

say if I have something like this.

```
 <root>
   <item>
     <title>First</title>
   </item>
   <item>
     <title>Second</title>
  </item>
  <somethingElse>
      <name>John </name>
  </somethingElse>
  <somethingElse>
      <name>John 2</name>
  </somethingElse>
 </root>
```

**Comment by chobo2** [comment]
Okay, I tried what you suggested on my real code and it was not working but the problem seems to be because I got other "elements"  what was not in my example.

Here is my new example

```
<root>
 <item>
      <name>test</name>
 </item>
<item>
    <name>test2</name>
</item>
<status>New</status>
</root>
```

```
[YAXSerializeAs("root")]
public class AllItems : List<Item>
{
  public string Status {get; set;}
}

public class item
{
    public string title { get; set; }
}

YAXSerilaizer ser = new YAXSerilaizer (typeof(Root));
var o = ser.Deserialize(xml);
```

**Reply by axunonb** [reply]
You should create the classes and decorate the members with attributes as needed.
Then create sample data and serialize it. The output structure should be the same as the XML you expect.
If you're so far, you're done and ready to deserialize "real" data.
Also having a brief look into the Wiki or even the unit tests can give more insights.
The most recent code sample doesn't compile, does it?

**Reply by chobo2** [reply]
Your code sample worked just fine but when I add extra elements like "status" to it then it fails with the "no element found".

**Reply by chobo2** [reply]
@axunonb 

I trying to modify the example though it is not giving me an error it is not giving me the result I am looking for

```
     static void Main(string[] args)
        {
            var data = new AllItems() { new() { title = "First" }, new() { title = "Second" } };
            data.External = "test";
            var s = new YAXSerializer<AllItems>();
            var xml = s.Serialize(data);
            /*<root>
              <External>test</External>
              <Item>
                <title>First</title>
              </Item>
              <Item>
                <title>Second</title>
              </Item>
            </root>*/
            var allItems = s.Deserialize(xml); // does not have the External property.
        }

        public class Item
        {
            public string title { get; set; }
        }

        [YAXSerializeAs("root")]
        public class AllItems : List<Item>
        {
            public string External { get; set; }
        }
```

**Comment by axunonb** [comment, marked as answer]
Aha, very good!

Well, `ICollection`s with additional fields or properties are a bit tricky.
If the deserializer sees an `ICollection`, it only uses the collection elements, but not any extra properties. Agree, this can't be expected, because the serializer uses any extra properties.

So, as a solution you could create a class `All` with the `string External` property, and a property `List<Item> AllItems`.
Decorate `AllItems` with the `YAXCollection` attribute as shown in the sample.
Then serialization and deserialization will bring the same results.

We'll add this use case to the docs.

```Csharp
public class Item
{
    public string title { get; set; }
}

[YAXSerializeAs("root")]
public class All
{
    public string External { get; set; }

    [YAXCollection(YAXCollectionSerializationTypes.RecursiveWithNoContainingElement, EachElementName = "item")]
    public List<Item> AllItems { get; set; }
}

[Test]
public void IList_Class_With_Extra_Properties()
{
    var data = new All { AllItems = new() { new() {title = "First" }, new() { title = "Second" }}, External = "external"};
    var s = new YAXSerializer<All>();
    var xml = s.Serialize(data);
    /*
<root>
  <External>external</External>
  <item>
    <title>First</title>
  </item>
  <item>
    <title>Second</title>
  </item>
</root>
     */
    var deserialized= s.Deserialize(xml);
    deserialized.Should().BeEquivalentTo(data); // test succeeds
}
```

**Reply by axunonb** [reply]
@304NotModified A bit tricky, isn't it?

**Reply by chobo2** [reply]
@axunonb  - That did it, yes, it is tricky.

To make it a bit trickier, my file that I am trying to parse has namespace, though the simple fields like "External" got picked up the collection one was always empty. I finally figured out I need to help and used "YAXNamespace"

Now I am off to my next problem. lol Not sure if I should create a new discussion though but basically, I have this

```
<root>
  <External>external</External>
  <item>
    <title>First</title>
    <notes>
         <div style="Padding-right: 0in;" xmlns="http:www.w3.org/1999/xhtml">
               <span style="font-weight: bold;">This is a note </span>
        </div>
         <div style="Padding-right: 0in;" xmlns="http:www.w3.org/1999/xhtml">
               <span style="font-weight: bold;">This is a note2 </span>
        </div>
         <div style="Padding-right: 0in;" xmlns="http:www.w3.org/1999/xhtml">
               <span style="font-weight: bold;">This is a note3 </span>
        </div>
   </notes>
  </item>
  <item>
    <title>Second</title>
    <notes>
            <div style="Padding-right: 0in;" xmlns="http:www.w3.org/1999/xhtml">
               <span style="font-weight: bold;">This is a note </span>
        </div>
         <div style="Padding-right: 0in;" xmlns="http:www.w3.org/1999/xhtml">
               <span style="font-weight: bold;">This is a note2 </span>
        </div>
         <div style="Padding-right: 0in;" xmlns="http:www.w3.org/1999/xhtml">
               <span style="font-weight: bold;">This is a note3 </span>
        </div>
   </notes>
  </item>
</root>

public class Item
{
    public string title { get; set; }
    public string notes {get; set;}
}

```

What I get as a result is: This is a note This is a note2 This is a note3

I don't get any of the div's or span tags

I thinking I have to make a Notes class, then a "div" class and then a "span" class and then maybe I would get the result.

However, I would like to avoid that as for this use came, I just need it as 1 block of text with the html tags in tack. I don't need to get the data line by line.

Result I would be hoping for is

```
            <div style="Padding-right: 0in;" xmlns="http:www.w3.org/1999/xhtml">
               <span style="font-weight: bold;">This is a note </span>
        </div>
         <div style="Padding-right: 0in;" xmlns="http:www.w3.org/1999/xhtml">
               <span style="font-weight: bold;">This is a note2 </span>
        </div>
         <div style="Padding-right: 0in;" xmlns="http:www.w3.org/1999/xhtml">
               <span style="font-weight: bold;">This is a note3 </span>
        </div>

```

**Reply by axunonb** [reply]
So actually this is much about parsing HTML where tools like `AngleSharp` are much better.
If you want to get all elements inside the `notes` node for further processing, you could consider to create an `ICustomSerializer`.

**Reply by chobo2** [reply]
> So actually this is much about parsing HTML where tools like `AngleSharp` are much better. If you want to get all elements inside the `notes` node for further processing, you could consider to create an `ICustomSerializer`.

yeah, I figure I would need another tool to parse those "divs" but right now I am not even getting access to any of those div tags so not sure if AngleSharp would help? 

Right now what I need is basically read all the xml and match them to the properties in my C# class which YaxLib does for me. There however is 1 field the 'notes' that contains html in it. I need to get everything in the notes fields including the html tags.

After that point I can use other tools or whatever to parse if I need more.

Why does it strip all the tags away instead of just plopping everything in the string?

**Reply by axunonb** [reply]
> strip all the tags away instead of just plopping everything in the string
>
😅 ... because this is not what XML is about. But when implementing a `ICustomSerializer` you could retrieve all descendants of the `notes` with no effort.

**Reply by chobo2** [reply]
@axunonb 

Well tell MS that lol, it's their files that I have to work with.

I looking at "CustomSerializationDemoClasses" but I am lost where to begin and how to implement it. Is there a more step by step tutorial on how to add ICustomSerializer?

**Reply by chobo2** [reply]
@axunonb 

Okay I think the best choice would be to use the example that put the property on the actual element.

I am looking at `public class CustomTitleSerializer : ICustomSerializer<string>`

I am looking at this method in the Serializer

```
    public string DeserializeFromElement(XElement element, ISerializationContext serializationContext)
    {
        return RetrieveValue(element.Value);
    }
```

I see if I do element.toString() I get everything back including the <note> elements. Like you said I should probably just get the descendants but not sure do I have to loop through the Next Node and build my own string?

**Reply by chobo2** [reply]
@axunonb 

Okay I think I got it, I am now returning a list of XNodes and using "element.DecendantNodes().ToList()" this way if I need to go line by line over I can.

I have a few xml elements that are blank in my document, If I want to insert a default value into those elements, do I have to make a custom Serializer? I see you have a default value for Error Handling if the element is missing, but my case it is there jsut nothing in it.

**Reply by axunonb** [reply]
Sorry for the late reply.
You could decorate properties, for which missing elements are allowed with `[YAXErrorIfMissed(YAXExceptionTypes.Ignore, DefaultValue = "what you like")]`. No need for a customer serializer as far as I understand your use case.
